task/control-plane-agent/src/inventory.rs

```diff
@@ -0,0 +1,201 @@
+// width is 16 bytes and the prefix is 4 bytes, allowing up to 999_999_999_999
+// devices to be listed.
+//
+// We tag this with `#[allow(dead_code)]` to prevent warnings about the contents
+// of this module not being used; the static assertion _is_ still checked.
+#[allow(dead_code)]
+mod max_num_devices {
+    use super::{SpComponent, VALIDATE_DEVICES};
+
+    // How many bytes are available for digits of a device index in base 10?
+    const DIGITS_AVAILABLE: usize =
+        SpComponent::MAX_ID_LENGTH - SpComponent::GENERIC_DEVICE_PREFIX.len();
+
+    // How many devices can we list given `DIGITS_AVAILABLE`?
+    const MAX_NUM_DEVICES: u64 = const_exp10(DIGITS_AVAILABLE);
+
+    // Statically assert that we have at most that many devices.
+    static_assertions::const_assert!(
+        VALIDATE_DEVICES.len() as u64 <= MAX_NUM_DEVICES
+    );
+
+    // Helper function: computes 10^n at compile time.
+    const fn const_exp10(mut n: usize) -> u64 {
+        let mut x = 1;
+        while n > 0 {
+            x *= 10;
+            n -= 1;
+        }
+        x
+    }
+}
+
+// We will spread the contents of `DEVICES` out over multiple packets to MGS;
+// however, we do _not_ currently handle the case where a single `DEVICES` entry
+// is too large to fit in a packet, even if it's the only device present in that
+// packet. Therefore, we assert at compile time via all the machinery below that
+// each entry of `DEVICES` is small enough that it will indeed fit in one packet
+// after being packed into a TLV triple.
+const ASSERT_EACH_DEVICE_FITS_IN_ONE_PACKET: () =
+    assert_each_device_tlv_fits_in_one_packet();
+
+const fn assert_device_tlv_fits_in_one_packet(
+    device: &'static str,
+    description: &'static str,
+) {
+    use gateway_messages::{
+        tlv, SerializedSize, MIN_SP_MESSAGE_TRAILING_TRAILING_DATA_LEN,
+    };
+
+    let encoded_len = tlv::tlv_len(
+        gateway_messages::DeviceDescriptionHeader::MAX_SIZE
+            + device.len()
+            + description.len(),
+    );
+
+    if encoded_len > MIN_SP_MESSAGE_TRAILING_TRAILING_DATA_LEN {
+        panic!(concat!(
+            "The device details (device and description) of at least one ",
+            "device in the current app.toml are too long to fit in a single ",
+            "TLV triple to send to MGS. Current Rust restrictions prevent us ",
+            "from being able to specific the specific device in this error ",
+            "message. Change this panic to `panic!(\"{{}}\", description)` ",
+            "and rebuild to see the description of the too-long device ",
+            "instead."
+        ))
+    }
+}
+
+const fn assert_each_device_tlv_fits_in_one_packet() {
+    let mut i = 0;
+    loop {
+        if i == VALIDATE_DEVICES.len() {
+            break;
+        }
+        assert_device_tlv_fits_in_one_packet(
+            VALIDATE_DEVICES[i].device,
+            VALIDATE_DEVICES[i].description,
+        );
+        i += 1;
+    }
```

Comment:
I'd vote to also check `DEVICES` here, since I was confused when initially reading this function.

This requires making `our_inventory::DEVICES` into `pub(super)`, but then you can delete the checks from `our_inventory` and have them all here, e.g. something like this:

```suggestion
    use our_inventory::DEVICES;
    let mut i = 0;
    loop {
        if i == DEVICES.len() {
            break;
        }
        assert_device_tlv_fits_in_one_packet(
            DEVICES[i].device,
            DEVICES[i].description,
        );
        i += 1;
    }
```